### ISSUE TITLE
SessionFlagManager starts session in CLI

### ISSUE THREAD

**Issue by umpirsky** [scheb/two-factor-bundle#63]
The issue is that in [SessionFlagManager.php#69](https://github.com/scheb/two-factor-bundle/blob/master/Security/TwoFactor/Session/SessionFlagManager.php#L69) the session will be started even in CLI.

It happened to me when calling `Symfony\Component\Security\Core\Authorization\AccessDecisionManager->decide($token, $role)` in CLI environment.

**Comment by scheb** [comment]
Might be solved with https://github.com/scheb/two-factor-bundle/commit/e9f3d5e309134b2ba9e6d8b08be5221e9fb33247, @umpirsky would you please check? I currently don't have a sample application running where I could try.

**Comment by umpirsky** [comment]
@scheb I doubt, but will check soon and post update here. Thanks!

**Comment by umpirsky** [comment]
@scheb No, it still happens with v2.6.

I prevented with:

```php
<?php

use Scheb\TwoFactorBundle\Security\TwoFactor\Session\SessionFlagManager as BaseSessionFlagManager;

class SessionFlagManager extends BaseSessionFlagManager
{
    public function isNotAuthenticated($provider, $token)
    {
        if ('cli' === php_sapi_name()) {
            return false;
        }

        return parent::isNotAuthenticated($provider, $token);
    }
}
```

Not very generic approach, any better ideas?

**Comment by zerkms** [comment]
@umpirsky aaaaaaaaaaaaaaaaaand this is another problem caused by the voter :-D

Now as you have disabled it (depending on how you did it though) I believe you don't need the hack above anymore.

**Comment by scheb** [comment]
@umpirsky Would you please check (backtrace) why/where the voter is triggered? In my demo application the CLI commands don't trigger the voter and therefore no session is created.

**Comment by umpirsky** [comment]
@scheb I'm calling `Symfony\Component\Security\Core\Authorization\AccessDecisionManager->decide($token, $role)` from CLI:

```
RuntimeException: Failed to start the session because headers have already been sent
1
File "/vendor/symfony/symfony/src/Symfony/Component/HttpFoundation/Session/Storage/NativeSessionStorage.php", line 144 in start
2
File "/vendor/symfony/symfony/src/Symfony/Component/HttpFoundation/Session/Storage/NativeSessionStorage.php", line 282 in getBag
3
File "/vendor/symfony/symfony/src/Symfony/Component/HttpFoundation/Session/Session.php", line 79 in has
4
File "/vendor/scheb/two-factor-bundle/Security/TwoFactor/Session/SessionFlagManager.php", line 69 in isNotAuthenticated
6
File "/vendor/scheb/two-factor-bundle/Security/TwoFactor/Voter.php", line 61 in vote
7
File "/vendor/symfony/symfony/src/Symfony/Component/Security/Core/Authorization/AccessDecisionManager.php", line 195 in decideUnanimous
8
File "/vendor/symfony/symfony/src/Symfony/Component/Security/Core/Authorization/AccessDecisionManager.php", line 66 in decide
9

```

**Comment by scheb** [comment]
Ok, but if you actually call the security/permission system, it's totally normal that the voter and dependent logic is called to determine permission. As a result a session is started, because it tries do get data from the session. So to me this doesn't look like an issue, but the result of intended behaviour.

**Comment by zerkms** [comment]
> but the result of intended behaviour.

The intention is to check whether one has permissions, not to start a session. Lack of session automatically means the lack of permission, there is no real reason to start session to find that.

So it's not an intended behaviour just not the best possible implementation.

**Comment by scheb** [comment]
Forgot to mention, the `isStarted` check is implemented since v2.6.2.

**Comment by umpirsky** [comment]
@scheb Sweet, thanks!
